src/styles/style-engine.ts

```diff
@@ -1,94 +1,61 @@
-import {View} from 'aurelia-templating';
-import {Origin} from 'aurelia-metadata';
-import {inject, Container} from 'aurelia-dependency-injection';
-import {StyleController} from './style-controller';
-import {DOM} from 'aurelia-pal';
-import {StyleFactory} from './style-factory';
-import {Themable} from './themable';
-import {camelCase} from 'aurelia-binding';
+import { inject } from 'aurelia-dependency-injection';
+import { StyleController } from './style-controller';
+import { UxTheme } from './ux-theme';
 
-@inject(Container)
+@inject(StyleController)
 export class StyleEngine {
-  private controllers: Map<any, StyleController> = new Map();
 
-  constructor(private container: Container) { }
-
-  public getThemeKeyForComponent(obj: any) {
-    return camelCase(Origin.get(obj.constructor).moduleMember + 'Theme');
-  }
-
-  public applyTheme(themable: Themable, theme: string | object | null) {
-    const themeKey = this.getThemeKeyForComponent(themable);
-    const currentController = (themable.view as any)[themeKey];
-    let bindingContext: any;
-    let newController: StyleController | undefined;
-
-    if (!theme) {
-      if (currentController !== currentController.factory.defaultController) {
-        currentController.unbind();
-        newController = currentController.factory.defaultController as StyleController;
-        (themable.view as any)[themeKey] = newController;
-        newController.bind(themable.view);
+  constructor(public styleController: StyleController) { }
+
+  /**
+   * Processes a UxTheme into the corresponding CSS Variables
+   * and applies them to the provided element.
+   *
+   * @param element Element to apply the processed UxTheme to.
+   * @param theme UxTheme to process.
+   */
+  public applyTheme(theme: UxTheme, element?: HTMLElement) {
+    if (element != null) {
+      for (const key in this.styleController.getThemeKeys(theme)) {
+        if (theme[key] != null) {
+          element.style.setProperty(`this.generateCssVariable(theme.themeKey, key, theme[key])`, theme[key]);

```

Comment:
I think there's more to it than this. For example, if it's a new theme instance, then we need to observe the properties on the new theme object for changes so that we can update the local css variables based on changes to the local theme.

Reply:
It's even trickier than that because it's possible that it's a shared theme object, but not the default object. So, we don't need to subscribe the second time, but there's an additional set of local css variable to update on changes. Does that make sense?